components/LineChart.js

```diff
@@ -0,0 +1,349 @@
+
+/**
+ * Given analytics stat, performs computations needed to plot a line chart.
+
+ * @param {array}  dataset List of data of the form {date, webVisit, lineVisit}
+ * @param {number} width   Width of plot area
+ * @param {number} height  Height of plot area
+
+ * @return {
+    xScale:      {d3.scaleTime}
+    yScaleWeb:   {d3.scaleLinear}
+    yScaleLine:  {d3.scaleLinear}
+    chatbotDots: {array}             List of corresponding positions for each chatbot entry
+    chatbotLine: {d3.line}
+    webDots:     {array}             List of corresponding positions for each web entry
+    webLine:     {d3.line}
+ }
+ */
+const computeChartData = (dataset, width, height) => {
+  const maxWebVisit = getMax(dataset, d => d.webVisit);
+  const maxLineVisit = getMax(dataset, d => d.lineVisit);
+
+  const xScale = scaleTime()
+    .domain([dataset[0].date, dataset[dataset.length - 1].date])
+    .range([0, width]);
+
+  const yScaleWeb = scaleLinear()
+    .domain([0, maxWebVisit])
+    .range([height, 0]);
+
+  const yScaleLine = scaleLinear()
+    .domain([0, maxLineVisit])
+    .range([height, 0]);
+
+  const webLine = line()
+    .x(d => xScale(d.date))
+    .y(d => yScaleWeb(d.webVisit))
+    .curve(curveMonotoneX);
+
+  const chatbotLine = line()
+    .x(d => xScale(d.date))
+    .y(d => yScaleLine(d.lineVisit))
+    .curve(curveMonotoneX);
+
+  const webDots = dataset.map(d => ({
+    value: d.webVisit,
+    date: d.date,
+    cx: xScale(d.date),
+    cy: yScaleWeb(d.webVisit),
+  }));
+  const chatbotDots = dataset.map(d => ({
+    value: d.lineVisit,
+    date: d.date,
+    cx: xScale(d.date),
+    cy: yScaleLine(d.lineVisit),
+  }));
+
+  return {
+    xScale,
+    yScaleWeb,
+    yScaleLine,
+    chatbotDots,
+    chatbotLine,
+    webDots,
+    webLine,
+  };
+};
+
+/* Renders gridline and axis tick label. */
+function TickGroup({ x, y, gridline, text }) {
+  return (
+    <g className="tickGroup" transform={`translate(${x},${y})`}>
+      <line className="gridline" x2={gridline.x2} y2={gridline.y2}></line>
+      <text fill="currentColor" x={text.x} y={text.y}>
+        {text.text}
+      </text>
+    </g>
+  );
+}
+/* Populates a list of tick values given a scale and number of ticks wanted. */
+const getTicks = (scale, roundFn, tickNum) => {
+  const [max, min] = scale.domain();
+  const diff = (max - min) / (tickNum - 1);
+  let ticks = [];
+  for (let i = 0; i < tickNum; i++) {
+    ticks.push(roundFn(max - diff * i));
+  }
+  return ticks;
+};
+
+const formatDate = (date, withYear = false) => {
+  const d = new Date(date);
+  const dateString = `${d.getMonth() + 1}/${d.getDate()}`; // MM/DD
+  if (withYear) {
+    return `${d.getFullYear()}/${dateString}`;
+  }
+  return dateString;
+};
+
+function plotTicks({ scale, x, y, text, roundFn, tickNum, gridline }) {
+  const ticks = getTicks(scale, roundFn, tickNum);
+  return ticks.map((tick, i) => (
+    <TickGroup
+      key={`${tick - 0}_${i}`}
```

Comment:
What does this `-0` for?

Reply:
`tick` could be a `number` or a `date`, `-0` remains the same if it's a number and converts `date` into epoch seconds otherwise.
it's not necessary, just makes the `key` look cleaner

Reply:
I see. We could use unary `+` to convert things to number, it's shorter ;)